From c32f5f9228809cd3fd2368bf566f1b6b0fa7c5b7 Mon Sep 17 00:00:00 2001
From: Michael Gill <michael.gill@xilinx.com>
Date: Tue, 12 Jul 2016 11:37:19 -0700
Subject: [PATCH 0901/1566] staging: apf: Fix address calculation for SG-DMA
 transfers

This patch comes from:
  https://github.com/Xilinx/linux-xlnx.git

If a buffer allocated with the use of sds_alloc is used in an
SG-DMA transfer such that only the end of the buffer is
transferred, the address calculation needed to skip the beginning
of the buffer was incorrect.  This patch fixes that issue.

Signed-off-by: Michael Gill <gill@xilinx.com>
(cherry picked from commit bbf9d396cae64edec762990701326d37c5ec90d1)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/staging/apf/xlnk.c |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/staging/apf/xlnk.c b/drivers/staging/apf/xlnk.c
index 135e0eb..83d5c80 100644
--- a/drivers/staging/apf/xlnk.c
+++ b/drivers/staging/apf/xlnk.c
@@ -1163,8 +1163,9 @@ static int xlnk_dmasubmit_ioctl(struct file *filp, unsigned int code,
 		void *kaddr = NULL;
 
 		if (buf_id) {
-			xlnk_intptr_type addr_delta = xlnk_phyaddr[buf_id] -
-				temp_args.dmasubmit.buf;
+			xlnk_intptr_type addr_delta =
+				temp_args.dmasubmit.buf -
+				xlnk_phyaddr[buf_id];
 			kaddr = (u8 *)(xlnk_bufpool[buf_id]) + addr_delta;
 		}
 		status = xdma_submit((struct xdma_chan *)
-- 
1.7.5.4

